File: tests/gen_genome_frags.cpp
// gen_genome_frags — Simulate aDNA reads as overlapping fragments of a reference genome.
//
// Unlike gen_synthetic (isolated molecules), reads here are drawn from a shared
// reference so longer reads overlap shorter ones and provide k-mers that extend
// beyond the shorter reads' termini.  This is the correct model for testing
// fqdup extend.
//
// Usage:
//   gen_genome_frags [options] > out.fq
//
// Options:
//   --genome-len N     Reference genome length    (default: 2000)
//   --n-reads   N      Number of reads            (default: 50000)
//   --min-len   N      Min fragment length        (default: 30)
//   --max-len   N      Max fragment length        (default: 80)
//   --dmax5     F      C→T amplitude at 5' end    (default: 0.0)
//   --dmax3     F      G→A amplitude at 3' end    (default: 0.0)
//   --lambda5   F      5' decay rate              (default: 0.35)
//   --lambda3   F      3' decay rate              (default: 0.35)
//   --qual      N      Phred base quality         (default: 37)
//   --seed      N      RNG seed                   (default: 42)

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <iostream>
#include <string>
#include <vector>

// ---------------------------------------------------------------------------
// PCG-32 PRNG (same as gen_synthetic)
// ---------------------------------------------------------------------------
struct PCG32 {
    uint64_t state, inc;
    explicit PCG32(uint64_t seed = 42, uint64_t stream = 1)
        : state(0), inc((stream << 1u) | 1u) {
        next(); state += seed; next();
    }
    uint32_t next() {
        uint64_t old = state;
        state = old * 6364136223846793005ULL + inc;
        uint32_t xsh = uint32_t(((old >> 18u) ^ old) >> 27u);
        uint32_t rot = uint32_t(old >> 59u);
        return (xsh >> rot) | (xsh << ((-rot) & 31u));
    }
    double uniform() { return (next() >> 8) * (1.0 / (1u << 24)); }
    // Uniform integer in [0, n)
    uint32_t range(uint32_t n) { return next() % n; }
};

static constexpr char BASES[4] = {'A', 'C', 'G', 'T'};
static constexpr char COMP[4]  = {'T', 'G', 'C', 'A'};  // complement of A,C,G,T

static int base_index(char c) {
    switch (c) {
        case 'A': case 'a': return 0;
        case 'C': case 'c': return 1;
        case 'G': case 'g': return 2;
        default:            return 3;
    }
}

static std::string revcomp(const std::string& s) {
    std::string rc(s.size(), 'N');
    for (int i = 0; i < (int)s.size(); ++i)
        rc[s.size() - 1 - i] = COMP[base_index(s[i])];
    return rc;
}

static std::string apply_damage(const std::string& seq,
                                double dmax5, double lambda5,
                                double dmax3, double lambda3,
                                PCG32& rng) {
    std::string s = seq;
    int L = (int)s.size();
    for (int p = 0; p < std::min(15, L); ++p) {
        double prob = dmax5 * std::exp(-lambda5 * p);
        if (s[p] == 'C' && rng.uniform() < prob) s[p] = 'T';
    }
    for (int p = 0; p < std::min(15, L); ++p) {
        int dist = L - 1 - p;
        if (dist < 0) break;
        double prob = dmax3 * std::exp(-lambda3 * p);
        if (s[dist] == 'G' && rng.uniform() < prob) s[dist] = 'A';
    }
    return s;
}

int main(int argc, char** argv) {
    int      genome_len = 2000;
    int      n_reads    = 50000;
    int      min_len    = 30;
    int      max_len    = 80;
    double   dmax5      = 0.0;
    double   dmax3      = 0.0;
    double   lambda5    = 0.35;
    double   lambda3    = 0.35;
    int      qual       = 37;
    uint64_t seed       = 42;

    for (int i = 1; i < argc; ++i) {
        auto d = [&]{ return std::stod(argv[++i]); };
        auto n = [&]{ return std::stoi(argv[++i]); };
        const char* a = argv[i];
        if      (!strcmp(a, "--genome-len")) genome_len = n();
        else if (!strcmp(a, "--n-reads"))    n_reads    = n();
        else if (!strcmp(a, "--min-len"))    min_len    = n();
        else if (!strcmp(a, "--max-len"))    max_len    = n();
        else if (!strcmp(a, "--dmax5"))      dmax5      = d();
        else if (!strcmp(a, "--dmax3"))      dmax3      = d();
        else if (!strcmp(a, "--lambda5"))    lambda5    = d();
        else if (!strcmp(a, "--lambda3"))    lambda3    = d();
        else if (!strcmp(a, "--qual"))       qual       = n();
        else if (!strcmp(a, "--seed"))       seed       = (uint64_t)n();
        else { std::cerr << "Unknown option: " << a << '\n'; return 1; }
    }

    std::cerr << "gen_genome_frags:"
              << " genome_len=" << genome_len
              << " n_reads="    << n_reads
              << " min_len="    << min_len
              << " max_len="    << max_len
              << " dmax5="      << dmax5
              << " dmax3="      << dmax3
              << '\n';

    PCG32 rng(seed);

    // Generate reference genome
    std::string genome(genome_len, 'N');
    for (char& c : genome) c = BASES[rng.range(4)];

    const char qual_char = (char)(qual + 33);

    for (int i = 0; i < n_reads; ++i) {
        int flen = min_len + (int)rng.range(max_len - min_len + 1);
        int start = (int)rng.range(genome_len - flen + 1);
        std::string seq = genome.substr(start, flen);

        // Randomly orient (simulate both strands)
        if (rng.uniform() < 0.5) seq = revcomp(seq);

        // Apply ancient DNA damage
        if (dmax5 > 0.0 || dmax3 > 0.0)
            seq = apply_damage(seq, dmax5, lambda5, dmax3, lambda3, rng);

        std::cout << '@' << 'r' << (i + 1) << '\n'
                  << seq << '\n'
                  << '+' << '\n'
                  << std::string(seq.size(), qual_char) << '\n';
    }

    return 0;
}

File: tests/gen_genome_frags.py
#!/usr/bin/env python3
"""
gen_genome_frags.py — Simulate aDNA reads as random fragments of a reference genome.

Unlike gen_synthetic (which generates isolated molecules), this tool samples reads
as overlapping fragments from a shared reference, so longer reads provide k-mers
that extend beyond shorter reads' termini.

This is the correct model for testing fqdup extend:
  - Each read = fragment [start, start+length) of the reference
  - C→T damage applied at the 5' end (position p from terminus)
  - G→A damage applied at the 3' end
  - Extension works because reads from different positions overlap each other

Usage:
  python3 gen_genome_frags.py [options] > out.fq

Options:
  --genome-len N     Reference genome length (default: 2000)
  --n-reads N        Number of reads to generate (default: 50000)
  --min-len N        Min fragment length (default: 30)
  --max-len N        Max fragment length (default: 80)
  --dmax5 F          C→T amplitude at 5' end (default: 0.0)
  --dmax3 F          G→A amplitude at 3' end (default: 0.0)
  --lambda5 F        5' decay rate (default: 0.35)
  --lambda3 F        3' decay rate (default: 0.35)
  --seed N           RNG seed (default: 42)
  --qual N           Base quality score (default: 37)
"""

import sys
import math
import random
import argparse

def parse_args():
    p = argparse.ArgumentParser()
    p.add_argument('--genome-len', type=int, default=2000)
    p.add_argument('--n-reads',    type=int, default=50000)
    p.add_argument('--min-len',    type=int, default=30)
    p.add_argument('--max-len',    type=int, default=80)
    p.add_argument('--dmax5',      type=float, default=0.0)
    p.add_argument('--dmax3',      type=float, default=0.0)
    p.add_argument('--lambda5',    type=float, default=0.35)
    p.add_argument('--lambda3',    type=float, default=0.35)
    p.add_argument('--seed',       type=int, default=42)
    p.add_argument('--qual',       type=int, default=37)
    return p.parse_args()

COMPLEMENT = str.maketrans('ACGTacgt', 'TGCAtgca')

def revcomp(s):
    return s.translate(COMPLEMENT)[::-1]

def apply_damage(seq, dmax5, dmax3, lam5, lam3, rng):
    """Apply C→T at 5' and G→A at 3' with exponential decay."""
    seq = list(seq)
    L = len(seq)
    for p in range(min(L, 15)):
        prob = dmax5 * math.exp(-lam5 * p)
        if seq[p] in ('C', 'c') and rng.random() < prob:
            seq[p] = 'T'
    for p in range(min(L, 15)):
        dist = L - 1 - p    # distance from 3' end
        if dist < 0:
            break
        prob = dmax3 * math.exp(-lam3 * p)
        if seq[dist] in ('G', 'g') and rng.random() < prob:
            seq[dist] = 'A'
    return ''.join(seq)

def main():
    args = parse_args()
    rng = random.Random(args.seed)

    # Generate reference genome
    bases = 'ACGT'
    genome = ''.join(rng.choice(bases) for _ in range(args.genome_len))

    sys.stderr.write(
        f"gen_genome_frags: genome_len={args.genome_len} n_reads={args.n_reads} "
        f"min_len={args.min_len} max_len={args.max_len} "
        f"dmax5={args.dmax5} dmax3={args.dmax3}\n"
    )

    qual_char = chr(args.qual + 33)
    out = sys.stdout

    for i in range(args.n_reads):
        # Sample fragment length and start position
        flen = rng.randint(args.min_len, args.max_len)
        start = rng.randint(0, args.genome_len - flen)
        seq = genome[start:start + flen]

        # Randomly orient (simulate both strands)
        if rng.random() < 0.5:
            seq = revcomp(seq)

        # Apply damage
        if args.dmax5 > 0 or args.dmax3 > 0:
            seq = apply_damage(seq, args.dmax5, args.dmax3,
                               args.lambda5, args.lambda3, rng)

        qual = qual_char * len(seq)
        out.write(f'@r{i+1}\n{seq}\n+\n{qual}\n')

if __name__ == '__main__':
    main()
